Add triggers documentation (#151)
memgraph/docs: docs/database-functionalities/triggers.md

---
id: triggers
title: Triggers
sidebar_label: Triggers
---

This article is a part of a series intended to showcase Memgraph's features
and bring the user up to speed on developing with Memgraph.

We highly recommend checking out the other articles from this series which
are listed in our [database functionalities section](./database-functionalities.md).

## Introduction

> Memgraph supports running openCypher statements after a certain event happens during transaction execution, i.e. triggers.
> 
> As user, you can create triggers, delete them, and print them out.

Reviewer: ```suggestion
Memgraph supports running openCypher statements after a certain event happens during transaction execution, i.e. triggers.

You can create triggers, delete them, and print them out.
```

All the triggers are persisted on the disk, so no information is lost between the runs.

## Creating a trigger

To create a new trigger, a query of the following format should be used:

```plaintext
CREATE TRIGGER trigger_name ( ON ( () | --> ) CREATE | UPDATE | DELETE )
( BEFORE | AFTER ) COMMIT
EXECUTE openCypherStatements
```

> At first, the query may seem complex, so let's talk about the different parts of it independently.

Reviewer: ```suggestion
The query may seem complex, so let's break it down.
```

### Trigger name

> Each created trigger should be defined with a name. The name needs to be unique globally.

Reviewer: ```suggestion
Each created trigger must have a globally unique name.
```

> E.g. two triggers for different event still can't have the same name.

Reviewer: ```suggestion
This implies that a pair of triggers for different events can't have the same name.
```

### Event type
Optionally, users can define for which event should the trigger execute its statements.
The event type is defined using the following part:

```plaintext
ON ( () | --> ) CREATE | UPDATE | DELETE
```

There are three main event types:
  - CREATE
  - UPDATE
  - DELETE

> For each of the event types, user can denote if he wants to execute trigger's statements just on the events

Reviewer 2: There are mixed usage of singular and plural form of user, so I would change everything to plural form, because that's more general in my opinion.
Sticking to singular form is also fine, then we are missing a few `a`/`the`.

> For each of the event types, user can denote if he wants to execute trigger's statements just on the events
> that happened on a vertex or just on an edge. Vertex is denoted using `()` and edges using `-->`.

Reviewer: ```suggestion
For each event type, users can specify whether to execute the trigger statements only on the events
that happened on a vertex, or on an edge. Vertices are denoted with `()`, and edges with `-->`.
```

Few examples would be:
`ON CREATE` - trigger the statements only if an object (vertex and/or edge) was created during the transaction execution.
`ON () UPDATE` - trigger the statements only if a vertex was updated (e.g. property was set on it) during the transaction execution.
`ON --> DELETE` - trigger the statements only if an edge was deleted during the transaction execution.

Each event comes with certain information that can be used in the openCypher statements the trigger executes. The information is contained in the
form of predefined vairables. (LINK TO THE PART ABOUT THE PREDEFINED VARIABLES)

If no event type is specified, the trigger executes its statements every time, and all the predefined variables can be used.

> ### Phase of executing statetements

Reviewer: ```suggestion
### Statement execution phase
```

> Trigger can execute its statements at specified phase, before or after comitting the transaction that triggered it.

Reviewer 2: I would also change `Trigger` -> `A trigger`, because that sounds more natural for me. Though grammarly agrees with me, I am still not sure. So feel free to keep this as is if you think this is better.

The same applies for below at line 68.

> Trigger can execute its statements at specified phase, before or after comitting the transaction that triggered it.

Reviewer: ```suggestion
A trigger can execute its statements at a specified phase, before or after committing the transaction that triggered it.
```

If `BEFORE COMMIT` option is used, the trigger will execute its statements as part of that transaction before it's committed.

> If `AFTER COMMIT` option is used, the trigger will execute its statements assynchrnously after that transaction is committed.

Reviewer: ```suggestion
If the `AFTER COMMIT` option is used, the trigger will execute its statements asynchronously after that transaction is committed.
```
